llm-examples/falcon/falcon-7b-instruct-inference.py

```diff
@@ -0,0 +1,349 @@
+        max_new_tokens = model_input.get("max_new_tokens", [100])[0]
+
+        # Build the prompt
+        prompt = self._build_prompt(prompt)
+
+        # Encode the input and generate prediction
+        encoded_input = self.tokenizer.encode(prompt, return_tensors='pt').to('cuda')
+        output = self.model.generate(encoded_input, do_sample=True, temperature=temperature, max_new_tokens=max_new_tokens)
+    
+        # Decode the prediction to text
+        generated_text = self.tokenizer.decode(output[0], skip_special_tokens=True)
+
+        # Removing the prompt from the generated text
+        prompt_length = len(self.tokenizer.encode(prompt, return_tensors='pt')[0])
+        generated_response = self.tokenizer.decode(output[0][prompt_length:], skip_special_tokens=True)
+
+        return [generated_response]
+
+# COMMAND ----------
+
+from mlflow.models.signature import ModelSignature
+from mlflow.types import DataType, Schema, ColSpec
+
+import pandas as pd
+
+# Define input and output schema
+input_schema = Schema([
+    ColSpec(DataType.string, "prompt"), 
+    ColSpec(DataType.double, "temperature"), 
+    ColSpec(DataType.long, "max_new_tokens")])
+output_schema = Schema([ColSpec(DataType.string)])
+signature = ModelSignature(inputs=input_schema, outputs=output_schema)
+
+# Define input example
+input_example=pd.DataFrame({
+            "prompt":["what is ML?"], 
+            "temperature": [0.5],
+            "max_new_tokens": [100]})
+
+# Log the model with its details such as artifacts, pip requirements and input example
+with mlflow.start_run() as run:  
+    mlflow.pyfunc.log_model(
+        "model",
+        python_model=Falcon(),
+        artifacts={'repository' : snapshot_location},
+        pip_requirements=["torch", "transformers", "accelerate", "einops","sentencepiece"],
+        input_example=input_example,
+        signature=signature,
+    )
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Register the model
+
+# COMMAND ----------
+
+# Register model in MLflow Model Registry
+result = mlflow.register_model(
+    "runs:/"+run.info.run_id+"/model",
+    name="falcon-7b-instruct",
+    await_registration_for=600,
+)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Load the model from model registry
+# MAGIC Assume that the below code is run separately or after the memory cache is cleared.
+# MAGIC
+# MAGIC You can clear the GPU memory by "Detach & re-attach" the notebook.
+
+# COMMAND ----------
+
+# MAGIC %pip install -q -U torch==2.0.1
+# MAGIC %pip install -q einops==0.6.1
+# MAGIC dbutils.library.restartPython()
+
+# COMMAND ----------
+
+import mlflow
+loaded_model = mlflow.pyfunc.load_model(f"models:/falcon-7b-instruct/latest")
+
+# COMMAND ----------
+
+import pandas as pd
+
+# Make a prediction using the loaded model
+input_example = pd.DataFrame({"prompt":["what is ML?"], "temperature": [0.1],"max_new_tokens": [100]})
+loaded_model.predict(input_example)
+
+# COMMAND ----------
```

Comment:
Add an example to run inference with spark UDF?

Reply:
I'm thinking we might not want to suggest this, because I found that it could easily run into OOM with `spark_udf` with such model size.

Reply:
Sure. Let's make a comments on this then.

Reply:
Q: Why do we get OOM? With spark each spark thread will be assigned one GPU, if the model can fit into one GPU, it should be fine, right? 
Did you run it on a single node cluster?